src/ucp/core/ucp_worker.c

```diff
@@ -2345,32 +2364,63 @@ static unsigned ucp_worker_discard_uct_ep_progress(void *arg)
     return 1;
 }
 
-static int ucp_worker_discard_remove_filter(const ucs_callbackq_elem_t *elem,
-                                            void *arg)
+static int
+ucp_worker_discard_remove_filter(const ucs_callbackq_elem_t *elem, void *arg)
 {
-    if ((elem->cb != ucp_worker_discard_uct_ep_destroy_progress) &&
-        (elem->cb != ucp_worker_discard_uct_ep_progress)) {
+    if ((elem->arg != arg) ||
+        ((elem->cb != ucp_worker_discard_uct_ep_destroy_progress) &&
+         (elem->cb != ucp_worker_discard_uct_ep_progress))) {
         return 0;
     }
 
-    ucp_worker_put_flush_req((ucp_request_t*)elem->arg);
+    ucp_worker_discard_uct_ep_complete((ucp_request_t*)elem->arg);
     return 1;
 }
 
-static void ucp_worker_discarded_uct_eps_cleanup(ucp_worker_h worker)
+static void
+ucp_worker_discard_uct_ep_pending_count_purged(uct_pending_req_t *self,
+                                               void *arg)
 {
+    ++(*((unsigned*)arg));
+}
+
+static void ucp_worker_discard_uct_ep_purge_reqs(ucp_worker_h worker)
+{
+    ucs_hlist_head_t *reqs = &worker->discard_uct_ep_reqs;
+    ucp_request_t *req;
+    unsigned num_purged;
     uct_ep_h uct_ep;
 
-    ucs_callbackq_remove_if(&worker->uct->progress_q,
-                            ucp_worker_discard_remove_filter, NULL);
+    while (!ucs_hlist_is_empty(reqs)) {
+        req        = ucs_hlist_head_elem(reqs, ucp_request_t, send.list);
+        uct_ep     = req->send.discard_uct_ep.uct_ep;
+        num_purged = 0;
+
+        uct_ep_pending_purge(uct_ep,
+                             ucp_worker_discard_uct_ep_pending_count_purged,
+                             &num_purged);
+        if (num_purged == 0) {
+            /* Either flush is in-progress (will be completed in UCT EP destroy)
+             * or some progress callback is scheduled (will be removed after UCT
+             * EP destroy) */
+            uct_ep_destroy(uct_ep);
```

Comment:
can move it out of branche (it is present in both branches)

Reply:
done